Tests/OpenCombineTests/PublisherTests/IgnoreOutputTests.swift

```diff
@@ -0,0 +1,159 @@
+//
+//  IgnoreOutputTests.swift
+//
+//  Created by Eric Patey on 16.08.20019.
+//
+
+import XCTest
+
+#if OPENCOMBINE_COMPATIBILITY_TEST
+import Combine
+#else
+import OpenCombine
+#endif
+
+@available(macOS 10.15, iOS 13.0, *)
+final class IgnoreOutputTests: XCTestCase {
+    static let allTests = [
+        ("testCompletionWithEmpty", testCompletionWithEmpty),
+        ("testCompletionWithValues", testCompletionWithValues),
+        ("testCompletionWithError", testCompletionWithError),
+        ("testDemand", testDemand),
+        ("testSendsSubcriptionDownstreamBeforeDemandUpstream",
+         testSendsSubcriptionDownstreamBeforeDemandUpstream),
+        ("testTestSuiteIncludesAllTests", testTestSuiteIncludesAllTests),
+    ]
+
+    func testCompletionWithEmpty() {
+        let subscription = CustomSubscription()
+        let publisher = CustomPublisher(subscription: subscription)
+        let ignoreOutputPublisher = publisher.ignoreOutput()
+        let tracking = TrackingSubscriberBase<Never, TestingError>(
+            receiveSubscription: { $0.request(.max(42)) }
+        )
+
+        XCTAssertEqual(tracking.history, [])
+
+        ignoreOutputPublisher.subscribe(tracking)
+        XCTAssertEqual(tracking.history, [.subscription("IgnoreOutput")])
+
+        publisher.send(completion: .finished)
+        XCTAssertEqual(tracking.history, [.subscription("IgnoreOutput"),
+                                          .completion(.finished)])
+    }
+
+    func testCompletionWithValues() {
+        let subscription = CustomSubscription()
+        let publisher = CustomPublisher(subscription: subscription)
+        let ignoreOutputPublisher = publisher.ignoreOutput()
+        let tracking = TrackingSubscriberBase<Never, TestingError>(
+            receiveSubscription: { $0.request(.max(42)) }
+        )
+
+        XCTAssertEqual(tracking.history, [])
+
+        ignoreOutputPublisher.subscribe(tracking)
+        XCTAssertEqual(tracking.history, [.subscription("IgnoreOutput")])
+
+        for _ in 0..<10 { _ = publisher.send(666) }
```

Comment:
Let's add `XCTAssertEqual` here instead of ignoring the return value to ensure that the underlying subscriber correctly handles demand:

```swift
for _ in 0 ..< 10 {
    XCTAssertEqual(publisher.send(666), .none)
}
```

Same applies in tests below.

I would also remove a loop, it seems unnecessary, but that's noncritical and up to you.

Reply:
I'll get rid of the loop.

To be honest, I am pretty confused about what the `Demand` result from `.receive(input:)` means semantically. The doc says:
> A `Demand` instance indicating how many more elements the subcriber expects to receive.

Why, then, would `IgnoreOutput` tell the upstream that it expects no more values? Maybe the wording that Apple chose isn't precise enough. When you look at the doc for the initial `Demand` given to the `Subscription`, it says:
> Tells a publisher that it may send more values to the subscriber.

So, in my mind, a `Subscription` must keep track of the current demand of the downstream `Subscriber` - initializing it with the value passed into `.request(demand:)` and updating it with the value returned by `.receive(input:)`.

So, given that logic, I would expect `IgnoreOutput` to return `.unlimited` rather than `.none` from `.receive(input:)`.

Reply:
We return `.none` because we requested `.unlimited` in the `receive(subscription:)` method, there's no point in requesting more as we've already requested the maximum possible value.

The `Demand` value returned from `receive(_:)` method may be useful if you request some finite value and then you realize you want more.

So yeah, probably the docs put it in a confusing way.

Reply:
Ah.  🧠💡I see. It's additive and not absolute. That's make so much more sense to me. D'oh.